app/src/main/java/io/homeassistant/companion/android/vehicle/DomainListScreen.kt

```diff
@@ -0,0 +1,84 @@
+package io.homeassistant.companion.android.vehicle
+
+import android.os.Build
+import androidx.annotation.RequiresApi
+import androidx.car.app.CarContext
+import androidx.car.app.Screen
+import androidx.car.app.model.Action
+import androidx.car.app.model.GridTemplate
+import androidx.car.app.model.Template
+import androidx.lifecycle.lifecycleScope
+import io.homeassistant.companion.android.BuildConfig
+import io.homeassistant.companion.android.common.R
+import io.homeassistant.companion.android.common.data.integration.Entity
+import io.homeassistant.companion.android.common.data.integration.IntegrationRepository
+import io.homeassistant.companion.android.common.data.integration.domain
+import io.homeassistant.companion.android.common.data.prefs.PrefsRepository
+import io.homeassistant.companion.android.common.data.servers.ServerManager
+import io.homeassistant.companion.android.util.vehicle.getDomainList
+import io.homeassistant.companion.android.util.vehicle.nativeModeActionStrip
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.launch
+
+@RequiresApi(Build.VERSION_CODES.O)
+class DomainListScreen(
+    carContext: CarContext,
+    val serverManager: ServerManager,
+    val integrationRepository: IntegrationRepository,
+    private val serverId: StateFlow<Int>,
+    private val allEntities: Flow<Map<String, Entity<*>>>,
+    private val prefsRepository: PrefsRepository
+) : Screen(carContext) {
+
+    private val domains = mutableSetOf<String>()
+
+    init {
+        lifecycleScope.launch {
+            allEntities.collect { entities ->
+                val newDomains = entities.values
+                    .map { it.domain }
+                    .distinct()
+                    .filter { it in MainVehicleScreen.SUPPORTED_DOMAINS }
+                    .toSet()
+                if (newDomains.size != domains.size || newDomains != domains) {
+                    domains.clear()
+                    domains.addAll(newDomains)
+                    invalidate()
+                }
+            }
+        }
+    }
+    override fun onGetTemplate(): Template {
+        val screen = MainVehicleScreen(
```

Comment:
You're only using the full screen here to get Automotive/driving optimized info. Can't we move those to the HaCarAppService to remove this need?

Reply:
i moved over as much as I could to keep the code easier so we dont have to call `MainVehicleScreen` like that not sure if this is desirable but I did it as best as I could since we only get some of this data in `Screen()`

Reply:
This does not appear to work correctly. Launching the app in the (very slow as always) Automotive API 32 emulator, I get the following crash:

```
2023-07-24 23:05:18.232 14581-14581 CarApp.H.Tem            com...mplates.host:renderer_service  E  Error: [type: null, cause: null, debug msg: java.lang.NullPointerException: null cannot be cast to non-null type android.car.drivingstate.CarUxRestrictionsManager
                                                                                                    	at io.homeassistant.companion.android.vehicle.MainVehicleScreen.registerAutomotiveRestrictionListener(MainVehicleScreen.kt:180)
                                                                                                    	at io.homeassistant.companion.android.vehicle.HaCarAppService$onCreateSession$1$2.onResume(HaCarAppService.kt:89)
                                                                                                    	at androidx.lifecycle.DefaultLifecycleObserverAdapter.onStateChanged(DefaultLifecycleObserverAdapter.kt:26)
                                                                                                    	at androidx.lifecycle.LifecycleRegistry$ObserverWithState.dispatchEvent(LifecycleRegistry.kt:314)
                                                                                                    	at androidx.lifecycle.LifecycleRegistry.forwardPass(LifecycleRegistry.kt:251)
                                                                                                    	at androidx.lifecycle.LifecycleRegistry.sync(LifecycleRegistry.kt:287)
                                                                                                    	at androidx.lifecycle.LifecycleRegistry.moveToState(LifecycleRegistry.kt:136)
                                                                                                    	at androidx.lifecycle.LifecycleRegistry.handleLifecycleEvent(LifecycleRegistry.kt:119)
                                                                                                    	at androidx.car.app.Session$LifecycleObserverImpl.onResume(Session.java:262)
                                                                                                    	at androidx.lifecycle.DefaultLifecycleObserverAdapter.onStateChanged(DefaultLifecycleObserverAdapter.kt:26)
                                                                                                    	at androidx.lifecycle.LifecycleRegistry$ObserverWithState.dispatchEvent(LifecycleRegistry.kt:314)
                                                                                                    	at androidx.lifecycle.LifecycleRegistry.forwardPass(LifecycleRegistry.kt:251)
                                                                                                    	at androidx.lifecycle.LifecycleRegistry.sync(LifecycleRegistry.kt:287)
                                                                                                    	at androidx.lifecycle.LifecycleRegistry.moveToState(LifecycleRegistry.kt:136)
                                                                                                    	at androidx.lifecycle.LifecycleRegistry.handleLifecycleEvent(LifecycleRegistry.kt:119)
                                                                                                    	at androidx.car.app.Session.handleLifecycleEvent(Session.java:187)
                                                                                                    	at androidx.car.app.CarAppBinder.lambda$onAppResume$2$androidx-car-app-CarAppBinder(CarAppBinder.java:170)
                                                                                                    	at androidx.car.app.CarAppBinder$$ExternalSyntheticLambda0.dispatch(Unknown Source:2)
                                                                                                    	at androidx.car.app.utils.RemoteUtils.lambda$dispatchCallFromHost$0(RemoteUtils.java:149)
                                                                                                    	at androidx.car.app.utils.RemoteUtils$$ExternalSyntheticLambda1.run(Unknown Source:6)
                                                                                                    	at androidx.car.app.utils.ThreadUtils.runOnMain(ThreadUtils.java:39)
                                                                                                    	at androidx.car.app.utils.RemoteUtils.dispatchCallFromHost(RemoteUtils.java:146)
                                                                                                    	at androidx.car.app.utils.RemoteUtils.lambda$dispatchCallFromHost$1(RemoteUtils.java:185)
                                                                                                    	at androidx.car.app.utils.RemoteUtils$$ExternalSyntheticLambda2.run(Unknown Source:8)
                                                                                                    	at android.os.Handler.handleCallback(Handler.java:938)
                                                                                                    	at android.os.Handler.dispatchMessage(Handler.java:99)
                                                                                                    	at android.os.Looper.loopOnce(Looper.java:201)
                                                                                                    	at android.os.Looper.loop(Looper.java:288)
                                                                                                    	at android.app.ActivityThread.main(ActivityThread.java:7870)
                                                                                                    	at java.lang.reflect.Method.invoke(Native Method)
                                                                                                    	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
                                                                                                    	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1003)
                                                                                                    ]
2023-07-24 23:05:18.244 15005-15005 AndroidRuntime          io....stant.companion.android.debug  E  FATAL EXCEPTION: main
                                                                                                    Process: io.homeassistant.companion.android.debug, PID: 15005
                                                                                                    java.lang.RuntimeException: java.lang.NullPointerException: null cannot be cast to non-null type android.car.drivingstate.CarUxRestrictionsManager
                                                                                                    	at androidx.car.app.utils.RemoteUtils.lambda$dispatchCallFromHost$0(RemoteUtils.java:154)
                                                                                                    	at androidx.car.app.utils.RemoteUtils$$ExternalSyntheticLambda1.run(Unknown Source:6)
                                                                                                    	at androidx.car.app.utils.ThreadUtils.runOnMain(ThreadUtils.java:39)
                                                                                                    	at androidx.car.app.utils.RemoteUtils.dispatchCallFromHost(RemoteUtils.java:146)
                                                                                                    	at androidx.car.app.utils.RemoteUtils.lambda$dispatchCallFromHost$1(RemoteUtils.java:185)
                                                                                                    	at androidx.car.app.utils.RemoteUtils$$ExternalSyntheticLambda2.run(Unknown Source:8)
                                                                                                    	at android.os.Handler.handleCallback(Handler.java:938)
                                                                                                    	at android.os.Handler.dispatchMessage(Handler.java:99)
                                                                                                    	at android.os.Looper.loopOnce(Looper.java:201)
                                                                                                    	at android.os.Looper.loop(Looper.java:288)
                                                                                                    	at android.app.ActivityThread.main(ActivityThread.java:7870)
                                                                                                    	at java.lang.reflect.Method.invoke(Native Method)
                                                                                                    	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
                                                                                                    	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1003)
                                                                                                    Caused by: java.lang.NullPointerException: null cannot be cast to non-null type android.car.drivingstate.CarUxRestrictionsManager
                                                                                                    	at io.homeassistant.companion.android.vehicle.MainVehicleScreen.registerAutomotiveRestrictionListener(MainVehicleScreen.kt:180)
                                                                                                    	at io.homeassistant.companion.android.vehicle.HaCarAppService$onCreateSession$1$2.onResume(HaCarAppService.kt:89)
                                                                                                    	at androidx.lifecycle.DefaultLifecycleObserverAdapter.onStateChanged(DefaultLifecycleObserverAdapter.kt:26)
                                                                                                    	at androidx.lifecycle.LifecycleRegistry$ObserverWithState.dispatchEvent(LifecycleRegistry.kt:314)
                                                                                                    	at androidx.lifecycle.LifecycleRegistry.forwardPass(LifecycleRegistry.kt:251)
                                                                                                    	at androidx.lifecycle.LifecycleRegistry.sync(LifecycleRegistry.kt:287)
                                                                                                    	at androidx.lifecycle.LifecycleRegistry.moveToState(LifecycleRegistry.kt:136)
                                                                                                    	at androidx.lifecycle.LifecycleRegistry.handleLifecycleEvent(LifecycleRegistry.kt:119)
                                                                                                    	at androidx.car.app.Session$LifecycleObserverImpl.onResume(Session.java:262)
                                                                                                    	at androidx.lifecycle.DefaultLifecycleObserverAdapter.onStateChanged(DefaultLifecycleObserverAdapter.kt:26)
                                                                                                    	at androidx.lifecycle.LifecycleRegistry$ObserverWithState.dispatchEvent(LifecycleRegistry.kt:314)
                                                                                                    	at androidx.lifecycle.LifecycleRegistry.forwardPass(LifecycleRegistry.kt:251)
                                                                                                    	at androidx.lifecycle.LifecycleRegistry.sync(LifecycleRegistry.kt:287)
                                                                                                    	at androidx.lifecycle.LifecycleRegistry.moveToState(LifecycleRegistry.kt:136)
                                                                                                    	at androidx.lifecycle.LifecycleRegistry.handleLifecycleEvent(LifecycleRegistry.kt:119)
                                                                                                    	at androidx.car.app.Session.handleLifecycleEvent(Session.java:187)
                                                                                                    	at androidx.car.app.CarAppBinder.lambda$onAppResume$2$androidx-car-app-CarAppBinder(CarAppBinder.java:170)
                                                                                                    	at androidx.car.app.CarAppBinder$$ExternalSyntheticLambda0.dispatch(Unknown Source:2)
                                                                                                    	at androidx.car.app.utils.RemoteUtils.lambda$dispatchCallFromHost$0(RemoteUtils.java:149)
                                                                                                    	at androidx.car.app.utils.RemoteUtils$$ExternalSyntheticLambda1.run(Unknown Source:6) 
                                                                                                    	at androidx.car.app.utils.ThreadUtils.runOnMain(ThreadUtils.java:39) 
                                                                                                    	at androidx.car.app.utils.RemoteUtils.dispatchCallFromHost(RemoteUtils.java:146) 
                                                                                                    	at androidx.car.app.utils.RemoteUtils.lambda$dispatchCallFromHost$1(RemoteUtils.java:185) 
                                                                                                    	at androidx.car.app.utils.RemoteUtils$$ExternalSyntheticLambda2.run(Unknown Source:8) 
                                                                                                    	at android.os.Handler.handleCallback(Handler.java:938) 
                                                                                                    	at android.os.Handler.dispatchMessage(Handler.java:99) 
                                                                                                    	at android.os.Looper.loopOnce(Looper.java:201) 
                                                                                                    	at android.os.Looper.loop(Looper.java:288) 
                                                                                                    	at android.app.ActivityThread.main(ActivityThread.java:7870) 
                                                                                                    	at java.lang.reflect.Method.invoke(Native Method) 
                                                                                                    	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548) 
                                                                                                    	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1003) 
```

Reply:
let me move it back then

Reply:
Editing the code like this resolves the crash, but that also proves `HaCarAppService()` is not an automatic singleton or anything like it. Let's keep the previous code for now instead of changing the architecture here.
```
val service = HaCarAppService()
service.car = Car.createCar(carContext)
service.carRestrictionManager =
    service.car?.getCarManager(Car.CAR_UX_RESTRICTION_SERVICE) as CarUxRestrictionsManager
val listener =
    CarUxRestrictionsManager.OnUxRestrictionsChangedListener { restrictions ->
        invalidate()
    }
service.carRestrictionManager?.registerListener(listener)
```

Reply:
done :)

Reply:
Should we duplicate the code perhaps, because the main vehicle screen will never be resumed and as a result `isDrivingOptimized` will always return false?

Reply:
wouldn't it already show `true` as `MainVehicleScreen` was already called beforehand?

Reply:
agreed we should run the check again here to ensure we have the most recent data

Reply:
> wouldn't it already show `true` as `MainVehicleScreen` was already called beforehand?

No because it is a new instance